go/client/cmd_ctl_start.go

```diff
@@ -4,33 +4,73 @@
 package client
 
 import (
+	"fmt"
+
 	"github.com/keybase/cli"
+	"github.com/keybase/client/go/install"
 	"github.com/keybase/client/go/libcmdline"
 	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol"
 )
 
 func NewCmdCtlStart(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Command {
 	return cli.Command{
 		Name:  "start",
 		Usage: "Start the background keybase service",
+		Flags: []cli.Flag{
+			cli.StringFlag{
+				Name:  "t, type",
+				Usage: fmt.Sprintf("Service type (%s, %s), default is %q", keybase1.ForkType_AUTO.String(), keybase1.ForkType_LAUNCHD.String(), defaultForkType.String()),
+			},
+		},
 		Action: func(c *cli.Context) {
-			cl.ChooseCommand(&CmdCtlStart{libkb.NewContextified(g)}, "start", c)
-			cl.SetForkCmd(libcmdline.ForceFork)
-			cl.SetNoStandalone()
+			cl.ChooseCommand(NewCmdCtlStartRunner(g), "start", c)
+			cl.SetForkCmd(libcmdline.NoFork)
+			cl.SetLogForward(libcmdline.LogForwardNone)
 		},
 	}
 }
 
 type CmdCtlStart struct {
 	libkb.Contextified
+	forkType keybase1.ForkType
+}
+
+func NewCmdCtlStartRunner(g *libkb.GlobalContext) *CmdCtlStart {
+	return &CmdCtlStart{Contextified: libkb.NewContextified(g)}
 }
 
 func (s *CmdCtlStart) ParseArgv(ctx *cli.Context) error {
+	forkTypeStr := ctx.String("type")
+	switch forkTypeStr {
+	case "", "launchd":
+		s.forkType = keybase1.ForkType_LAUNCHD
+	case "auto":
+		s.forkType = keybase1.ForkType_AUTO
+	default:
+		return fmt.Errorf("Invalid type: %s", forkTypeStr)
+	}
 	return nil
 }
 
-func (s *CmdCtlStart) Run() (err error) {
-	return nil
+func (s *CmdCtlStart) runAutoFork() error {
+	_, err := AutoForkServer(s.G(), s.G().Env.GetCommandLine())
+	return err
+}
+
+func (s *CmdCtlStart) runLaunchd() error {
+	return StartLaunchdService(s.G(), string(install.AppServiceLabel), true)
+}
+
+func (s *CmdCtlStart) Run() error {
+	switch s.forkType {
+	case keybase1.ForkType_LAUNCHD:
+		return s.runLaunchd()
+	case keybase1.ForkType_AUTO:
+		return s.runAutoFork()
+	default:
+		return fmt.Errorf("Unsupported service type")
+	}
```

Comment:
I worry the leakage will force us to create more and more branches like this. (And [this](https://github.com/keybase/client/blob/47f74b2764d95532b7cb104caf94aa76077c5464/go/client/versionfix.go#L123).) Is there a way to bottleneck more of this code through some platform-independent functions?